include/albatross/src/graph/minimum_spanning_tree.hpp

```diff
@@ -173,6 +173,92 @@ Graph<VertexType> maximum_spanning_tree(const Graph<VertexType> &graph) {
   return output;
 }
 
+// A helper struct to track if a closed loop would be formed by a proposed edge
+// during Kruskal's algorithm. This struct associates each node with a tree,
+// initialized to each node in a unique tree. When a new edge is added, the
+// trees associated with the edge's nodes are merged into one tree. If a
+// proposed edge already has both nodes in the same tree, a closed loop would be
+// formed, and the edge can be rejected.
+template <typename VertexType> struct KruskalNodeTracker {
+
+  struct VertexWithTreeID {
+    VertexWithTreeID(const VertexType &v_, const size_t &tree_)
+        : v(v_), tree(tree_){};
+    VertexType v;
+    size_t tree;
+  };
+
+  KruskalNodeTracker(const std::set<VertexType> &vertices) : vertices_() {
+    size_t tree_id = 0;
+    for (const auto &v : vertices) {
+      vertices_.emplace_back(v, tree_id);
+      tree_id++;
+    }
+  };
+
+  bool in_same_tree_else_add_edge(const VertexType &u, const VertexType &v) {
+    size_t index_u, index_v; // can remove this lookup if vertices are replaced
+                             // with consecutive 0-indexed IDs. This function
+                             // serves two roles to avoid repeated lookups
+    bool found_u = false;
```

Comment:
nit: I think if I'd implemented this I would have tried to use `std::find_if`, something like:
```
auto find_or_assert = [&](const VertexType &x) {
  auto is_x = [&x](const auto &p) { return p.v == x; };
  const auto iter = std::find_if(vertices_.begin(), vertices_.end(), is_x);
  assert(iter != vertices_.end());
  return *iter;
}
const VertexWithTreeID u_and_tree = find_or_assert(u);
const VertexWithTreeID v_and_tree = find_or_assert(v);
```
Then subsequent steps don't need to do lookups:
```
if (u_and_tree.tree == v_and_tree.tree)
  ...
```
but while that's probably more c++ style, I'm not really sure if it's more clear/efficient.

Reply:
It's likely more efficient unless the compiler optimizes away the incrementing counter when using the iterator. The downside is that it performs a separate lookup for each vertex when both can be looked for on the same iteration. But I could see a variation where both are found in one pass, but this is a premature optimization if a better implementation would be to replace vertices with IDs.
```
auto find_or_assert = [&](const VertexType &x, const VertexType &x) {
  auto is_x_or_y = [&x](const auto &p) { return p.v == x || p.v == y; };
  const auto iter_1 = std::find_if(vertices_.begin(), vertices_.end(), is_x_or_y);
  const auto iter_2 = std::find_if(iter_1++, vertices_.end(), is_x_or_y);
  return std::pair(*iter_1,*iter_2)
}
```
I'll try your suggestion.